Deserialization: floating point bounds check does not permit messages with NaN

## Bug report

**Required Info:**

- Operating System: Linux 
- Installation type: Source, master
- Version or commit hash: 3dc587255effce330db05e045a5cead78df28339
- DDS implementation: Fast-RTPS
- Client library (if applicable): rclpy

#### Steps to reproduce issue
Running deserialize_message(serialized_message: bytes, message_type) on a message containing a field set to NaN triggers the bounds check added in #128.

assert value >= -3.402823e+38 and value <= 3.402823e+38,
AssertionError: The 'current' field must be a float in [-3.402823e+38, 3.402823e+38]

#### Expected behavior
The NaN value should be converted to a NaN in python.

The [BatteryState](https://github.com/ros2/common_interfaces/blob/b7e2ad503e1648468b1dc766c43aac12cc9aee5a/sensor_msgs/msg/BatteryState.msg) message, for example, uses NaNs to flag values that are not measured.

#### Actual behavior
The conversion triggers an assertion, and fails to deserialize.

#### Additional information

I'm able to reproduce. The check happens on the setter for a message field of types `float`/`float32` and `double`/`float64` (.idl/.msg). It looks like the check's purpose is to make sure the value can be represented by the storage type. While the ticket mentions NaN, the logic also fails for positive and negative `inf`.

```python3
>>> from sensor_msgs.msg import BatteryState
>>> b = BatteryState()
>>> b.temperature = float('nan')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sloretz/ws/ros2/install/sensor_msgs/local/lib/python3.10/dist-packages/sensor_msgs/msg/_battery_state.py", line 424, in temperature
    assert value >= -3.402823e+38 and value <= 3.402823e+38, \
AssertionError: The 'temperature' field must be a float in [-3.402823e+38, 3.402823e+38]
>>> b.temperature = float('inf')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sloretz/ws/ros2/install/sensor_msgs/local/lib/python3.10/dist-packages/sensor_msgs/msg/_battery_state.py", line 424, in temperature
    assert value >= -3.402823e+38 and value <= 3.402823e+38, \
AssertionError: The 'temperature' field must be a float in [-3.402823e+38, 3.402823e+38]
>>> from sensor_msgs.msg import RelativeHumidity
>>> r = RelativeHumidity()
>>> r.variance =  float('nan')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sloretz/ws/ros2/install/sensor_msgs/local/lib/python3.10/dist-packages/sensor_msgs/msg/_relative_humidity.py", line 170, in variance
    assert value >= -1.7976931348623157e+308 and value <= 1.7976931348623157e+308, \
AssertionError: The 'variance' field must be a double in [-1.7976931348623157e+308, 1.7976931348623157e+308]
>>> r.variance =  float('inf')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sloretz/ws/ros2/install/sensor_msgs/local/lib/python3.10/dist-packages/sensor_msgs/msg/_relative_humidity.py", line 170, in variance
    assert value >= -1.7976931348623157e+308 and value <= 1.7976931348623157e+308, \
AssertionError: The 'variance' field must be a double in [-1.7976931348623157e+308, 1.7976931348623157e+308]

```

--- a/rosidl_generator_py/CMakeLists.txt
+++ b/rosidl_generator_py/CMakeLists.txt
@@ -6,7 +6,6 @@ find_package(ament_cmake REQUIRED)
 
 ament_export_dependencies(ament_cmake)
 ament_export_dependencies(rmw)
-ament_export_dependencies(rosidl_cmake)
 
 ament_index_register_resource("rosidl_generator_packages")
 

--- a/rosidl_generator_py/package.xml
+++ b/rosidl_generator_py/package.xml
@@ -21,8 +21,8 @@
   <buildtool_export_depend>ament_cmake</buildtool_export_depend>
   <buildtool_export_depend>ament_index_python</buildtool_export_depend>
   <buildtool_export_depend>python_cmake_module</buildtool_export_depend>
-  <buildtool_export_depend>rosidl_cmake</buildtool_export_depend>
   <buildtool_export_depend>rosidl_generator_c</buildtool_export_depend>
+  <buildtool_export_depend>rosidl_pycommon</buildtool_export_depend>
   <buildtool_export_depend>rosidl_typesupport_c</buildtool_export_depend>
   <buildtool_export_depend>rosidl_typesupport_interface</buildtool_export_depend>
 

--- a/rosidl_generator_py/resource/_action.py.em
+++ b/rosidl_generator_py/resource/_action.py.em
@@ -1,6 +1,6 @@
 @# Included from rosidl_generator_py/resource/_idl.py.em
 @{
-from rosidl_cmake import convert_camel_case_to_lower_case_underscore
+from rosidl_pycommon import convert_camel_case_to_lower_case_underscore
 
 action_name = '_' + convert_camel_case_to_lower_case_underscore(action.namespaced_type.name)
 module_name = '_' + convert_camel_case_to_lower_case_underscore(interface_path.stem)

--- a/rosidl_generator_py/resource/_action_pkg_typesupport_entry_point.c.em
+++ b/rosidl_generator_py/resource/_action_pkg_typesupport_entry_point.c.em
@@ -55,7 +55,7 @@ TEMPLATE(
 }@
 
 @{
-from rosidl_cmake import convert_camel_case_to_lower_case_underscore
+from rosidl_pycommon import convert_camel_case_to_lower_case_underscore
 type_name = convert_camel_case_to_lower_case_underscore(action.namespaced_type.name)
 function_name = 'type_support'
 }@

--- a/rosidl_generator_py/resource/_idl_pkg_typesupport_entry_point.c.em
+++ b/rosidl_generator_py/resource/_idl_pkg_typesupport_entry_point.c.em
@@ -14,7 +14,7 @@
 @# Handle messages
 @#######################################################################
 @{
-from rosidl_cmake import convert_camel_case_to_lower_case_underscore
+from rosidl_pycommon import convert_camel_case_to_lower_case_underscore
 from rosidl_parser.definition import Message
 
 include_directives = set()

--- a/rosidl_generator_py/resource/_msg.py.em
+++ b/rosidl_generator_py/resource/_msg.py.em
@@ -37,7 +37,7 @@ if message.structure.members:
         'import rosidl_parser.definition', [])  # used for SLOT_TYPES
 for member in message.structure.members:
     type_ = member.type
-    if isinstance(type_, AbstractNestedType): 
+    if isinstance(type_, AbstractNestedType):
         type_ = type_.value_type
     if member.name != EMPTY_STRUCTURE_REQUIRED_MEMBER_NAME:
         imports.setdefault(

--- a/rosidl_generator_py/resource/_msg.py.em
+++ b/rosidl_generator_py/resource/_msg.py.em
@@ -1,7 +1,7 @@
 @# Included from rosidl_generator_py/resource/_idl.py.em
 @{
 
-from rosidl_cmake import convert_camel_case_to_lower_case_underscore
+from rosidl_pycommon import convert_camel_case_to_lower_case_underscore
 from rosidl_generator_py.generate_py_impl import constant_value_to_py
 from rosidl_generator_py.generate_py_impl import get_python_type
 from rosidl_generator_py.generate_py_impl import SPECIAL_NESTED_BASIC_TYPES

--- a/rosidl_generator_py/resource/_msg_pkg_typesupport_entry_point.c.em
+++ b/rosidl_generator_py/resource/_msg_pkg_typesupport_entry_point.c.em
@@ -1,6 +1,6 @@
 @# Included from rosidl_generator_py/resource/_idl_pkg_typesupport_entry_point.c.em
 @{
-from rosidl_cmake import convert_camel_case_to_lower_case_underscore
+from rosidl_pycommon import convert_camel_case_to_lower_case_underscore
 
 include_parts = idl_type.namespaces + [
     'detail', convert_camel_case_to_lower_case_underscore(idl_type.name)]

--- a/rosidl_generator_py/resource/_msg_support.c.em
+++ b/rosidl_generator_py/resource/_msg_support.c.em
@@ -683,7 +683,7 @@ nested_type = '__'.join(type_.namespaced_name())
       (void)rc;
       assert(rc == 0);
 @[    elif isinstance(member.type.value_type, AbstractString)]@
-      PyObject * decoded_item = PyUnicode_DecodeUTF8(src[i].data, strlen(src[i].data), "strict");
+      PyObject * decoded_item = PyUnicode_DecodeUTF8(src[i].data, strlen(src[i].data), "replace");
       if (!decoded_item) {
         return NULL;
       }
@@ -749,7 +749,7 @@ nested_type = '__'.join(type_.namespaced_name())
     field = PyUnicode_DecodeUTF8(
       ros_message->@(member.name).data,
       strlen(ros_message->@(member.name).data),
-      "strict");
+      "replace");
     if (!field) {
       return NULL;
     }

--- a/rosidl_generator_py/resource/_msg_support.c.em
+++ b/rosidl_generator_py/resource/_msg_support.c.em
@@ -1,7 +1,7 @@
 @# Included from rosidl_generator_py/resource/_idl_support.c.em
 #define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
 @{
-from rosidl_cmake import convert_camel_case_to_lower_case_underscore
+from rosidl_pycommon import convert_camel_case_to_lower_case_underscore
 from rosidl_generator_py.generate_py_impl import SPECIAL_NESTED_BASIC_TYPES
 from rosidl_parser.definition import AbstractNestedType
 from rosidl_parser.definition import AbstractSequence

--- a/rosidl_generator_py/resource/_srv.py.em
+++ b/rosidl_generator_py/resource/_srv.py.em
@@ -1,6 +1,6 @@
 @# Included from rosidl_generator_py/resource/_idl.py.em
 @{
-from rosidl_cmake import convert_camel_case_to_lower_case_underscore
+from rosidl_pycommon import convert_camel_case_to_lower_case_underscore
 
 service_name = '_' + convert_camel_case_to_lower_case_underscore(service.namespaced_type.name)
 module_name = '_' + convert_camel_case_to_lower_case_underscore(interface_path.stem)

--- a/rosidl_generator_py/resource/_srv_pkg_typesupport_entry_point.c.em
+++ b/rosidl_generator_py/resource/_srv_pkg_typesupport_entry_point.c.em
@@ -18,7 +18,7 @@ TEMPLATE(
 }@
 @
 @{
-from rosidl_cmake import convert_camel_case_to_lower_case_underscore
+from rosidl_pycommon import convert_camel_case_to_lower_case_underscore
 type_name = convert_camel_case_to_lower_case_underscore(service.namespaced_type.name)
 function_name = 'type_support'
 }@

--- a/rosidl_generator_py/rosidl_generator_py/generate_py_impl.py
+++ b/rosidl_generator_py/rosidl_generator_py/generate_py_impl.py
@@ -18,11 +18,6 @@ import os
 import pathlib
 import sys
 
-from rosidl_cmake import convert_camel_case_to_lower_case_underscore
-from rosidl_cmake import expand_template
-from rosidl_cmake import generate_files
-from rosidl_cmake import get_newest_modification_time
-from rosidl_cmake import read_generator_arguments
 from rosidl_parser.definition import AbstractGenericString
 from rosidl_parser.definition import AbstractNestedType
 from rosidl_parser.definition import AbstractSequence
@@ -38,6 +33,11 @@ from rosidl_parser.definition import Message
 from rosidl_parser.definition import NamespacedType
 from rosidl_parser.definition import Service
 from rosidl_parser.parser import parse_idl_file
+from rosidl_pycommon import convert_camel_case_to_lower_case_underscore
+from rosidl_pycommon import expand_template
+from rosidl_pycommon import generate_files
+from rosidl_pycommon import get_newest_modification_time
+from rosidl_pycommon import read_generator_arguments
 
 SPECIAL_NESTED_BASIC_TYPES = {
     'float': {'dtype': 'numpy.float32', 'type_code': 'f'},

--- a/rosidl_generator_py/test/test_interfaces.py
+++ b/rosidl_generator_py/test/test_interfaces.py
@@ -487,14 +487,20 @@ def test_arrays():
         setattr(msg, 'uint64_values', [-1, 1, 2])
     with pytest.raises(AssertionError):
         float32_ieee_max_next = numpy.nextafter(3.402823466e+38, math.inf)
-        setattr(msg, 'float32_values', [-float32_ieee_max_next, 0.0, float32_ieee_max_next])
+        setattr(msg, 'float32_values', [-float32_ieee_max_next, 0.0])
+    with pytest.raises(AssertionError):
+        float32_ieee_max_next = numpy.nextafter(3.402823466e+38, math.inf)
+        setattr(msg, 'float32_values', [0.0, float32_ieee_max_next])
 
     # If target system is IEEE 754 compliant, the next number is rounded to inf.
     # Only perform this check on non-compliant systems.
     if sys.float_info.max > 1.7976931348623157e+308:
         with pytest.raises(AssertionError):
             float64_ieee_max_next = numpy.nextafter(1.7976931348623157e+308, math.inf)
-            setattr(msg, 'float64_values', [-float64_ieee_max_next, 0.0, float64_ieee_max_next])
+            setattr(msg, 'float64_values', [-float64_ieee_max_next, 0.0])
+        with pytest.raises(AssertionError):
+            float64_ieee_max_next = numpy.nextafter(1.7976931348623157e+308, math.inf)
+            setattr(msg, 'float64_values', [-float64_ieee_max_next, 0.0])
 
     # NaN
     arr_of_float32_with_nan = numpy.array([-1.33, math.nan, 1.33], dtype=numpy.float32)
@@ -724,7 +730,10 @@ def test_bounded_sequences():
         setattr(msg, 'uint64_values', [-1, 1, 2])
     with pytest.raises(AssertionError):
         float32_ieee_max_next = numpy.nextafter(3.402823466e+38, math.inf)
-        setattr(msg, 'float32_values', [-float32_ieee_max_next, 0.0, float32_ieee_max_next])
+        setattr(msg, 'float32_values', [-float32_ieee_max_next, 0.0])
+    with pytest.raises(AssertionError):
+        float32_ieee_max_next = numpy.nextafter(3.402823466e+38, math.inf)
+        setattr(msg, 'float32_values', [0.0, float32_ieee_max_next])
 
     # If target system is IEEE 754 compliant, the next number is rounded to inf.
     # Only perform this check on non-compliant systems.

